add new header to expose topology get APIs

tpatki opened the pull request:
Refactors Stephanie's PR #268 by exposing a `hwloc` struct variable to the users in `variorum_topology.h`, adding `init/destroy` functions, and refactoring `config_architecture.c` to use `variorum_topology.h`. 

- adds variorum_topology.h, which exposes APIs to get basic topology info
including number of sockets, cores, and threads
- installs this header alongside variorum.h
- Resolves https://github.com/LLNL/variorum/issues/266

Some advantages include:

- no code duplication
- users can use the same hwloc topology object to do other complex tasks in their code if needed (by including `hwloc.h` of course, and calling functions there in addition to variorum_topology API for `num_sockets` `num_cores` etc). 

Questions:
- Unclear if adding the struct to variorum_topology.h creates other problems, unsure if this needs to be static.

Testing:
- Works on Lassen. Tested with the `variorum_print_power`, `variorum_print_power_limit` and `variorum_get_node_power_json` APIs, as well as the following test:

```
#include <stdio.h>
#include <stdlib.h>

#include <variorum.h>
#include <variorum_topology.h>

int main(int argc, char **argv)
{
    printf("Num sockets %d\n", variorum_get_num_sockets());
    printf("Num cores %d\n", variorum_get_num_cores());
    printf("Num threads %d\n", variorum_get_num_threads());
}

```
Built as follows:
```
$ gcc test.c -I/g/g90/patki1/src/var-clean/variorum/install-topo-test/include/ -L/g/g90/patki1/src/var-clean/variorum/install-topo-test/lib/ -lvariorum

$ ./a.out 
Num sockets 2
Num cores 40
Num threads 160
```

ToDo testing:
- [x] Comus (Broadwell), 
- [x] Thompson (Haswell), 
- [x] Alehouse (Power9)
- [x] Rhetoric (Skylake, single socket)
- [x] ARM Juno r2
- [x] Lassen with the CUDA host-config file build for NVML testing (different than alehouse).
- [x] Cleanup print statements
- [x] Add example .

slabasan committed:
add new header to expose topology get APIs

- adds variorum_topology.h, which exposes APIs to get basic topology info
  including as number of sockets, cores, and threads
- installs this header alongside variorum.h

tpatki committed:
Another attempt to refactor HWLOC code and expose get_topology style APIs without code duplication.

tpatki committed:
Refactor config_architecture.c to avoid code duplication... code needs minor cleanup

tpatki committed:
Cleanup. Fix build error.

tpatki committed:
Build complains about gethostname header...

slabasan commented:
Should we aim to have a single hwloc topology object in all of variorum? With this structure, I believe we will be creating multiple topology objects, one for each call to the new variorum topology get functions, and then for any front-facing variorum API.

tpatki commented:
@slabasan -- Yes -- we need a single topology object for all of variorum -- which is what I was trying to do by moving the `topology` struct to the header, and using static. For some reason that didn't work with the build system (got errors on some compilers and not on others) -- not sure how to fix that. Seemed quite simple and obvious but not sure what I was missing on Friday (may just have been tired to think).

slabasan committed:
code is working

- move topology declaration into variorum_topology.h
- make topology object static in variorum_topology.h
- comment out topology destroy for now...makes things more complex with wanting
  to init topology object 1x across all functions
- use static init variable in variorum_init_topology to only create topology
  object once
- use extern in variorum_topology.c to resolve gcc7 and gcc11 warnings about
  varible defined but not used

tpatki committed:
Try to use extern correctly. Header says extern, one of the C files defines it

tpatki committed:
Cleanup code. remove debug print statements, and remove references to hwloc_destroy

tpatki committed:
Add an example for usage of variorum_topology.h

slabasan committed:
fix astyle errors

slabasan committed:
remove unused variables
